csrc/device_lower/pass/circular_buffer.cpp

```diff
@@ -215,6 +217,732 @@ class CircularBufferLoopCloner : public kir::IrVisitor {
+  // are added sequentially. For TmaCircularBufferLoop, we have an mbarrier for
+  // each Tensorview and each circular buffer stage, but not for each individual
+  // TMA load operation. If there are serial IterDomains to the right of the
+  // computeAt position, nvfuser will generate a for-loop to launch multiple TMA
+  // load operations.
+  //
+  // When we encounter a CpAsyncBulk load expression, we create a mbarrier_wait
+  // for the main and epilogue loops and a arriveExpectTx for prologue and main
+  // loops. The expected_tx for arriveExpectTx is the cumulative transaction
+  // size for all TMA load operations for the TensorView. Next, we generate the
+  // nested for-loops for the serial IterDomains, but do not add them to the
+  // cloned circular buffer loop immediately. Once the cloned circular buffer
+  // loop is the only loop in the stack, add the arriveExpectTx and arrive
+  // expressions, then the nested for-loop structure calling the TMA load
+  // operations, and finally the mbarrier_wait.
+  void processForLoop(ForLoop* cloned_loop) final {
+    // Skip if there is not an active for-loop structure
+    if (for_loop_stack_.empty()) {
+      return;
+    }
+
+    if (!cloned_loop->body().empty()) {
+      // mbarrier_arrive_tx_ is active when we encounter a cpAsyncBulk load
+      // operation on a circular buffer TensorView in IrVisitor. A single
+      // mbarrier_arrive_tx is active for each TensorView.
+      if (mbarrier_arrive_tx_ == nullptr || for_loop_stack_.size() > 1) {
+        // Add cloned for_loop when mbarrier_arrive_tx_ is not active or
+        // we are within a nested for-loop structure
+        for_loop_stack_.back()->body().push_back(cloned_loop);
+      } else {
+        // mbarrier::arriveExpectTx and TMA load operations occur in prologue
+        // and main loops.
+        NVF_ERROR(for_loop_stack_.front() == cloned_top_level_loop_);
+        addTmaLoadBlock(cloned_loop);
+      }
+    }
+
+    // mbarrier::wait occurs in Main and Epilogue loops.
+    if (mbarrier_wait_ != nullptr && for_loop_stack_.size() == 1) {
+      NVF_ERROR(for_loop_stack_.back() == cloned_top_level_loop_);
+      cloned_top_level_loop_->body().push_back(mbarrier_wait_);
+      mbarrier_wait_ = nullptr;
+    }
+  }
+
+  void processExpr(Expr* expr) final {
+    // A mbarrier token exists if the TensorView output for cpAsyncBulk load has
+    // circular buffering depth > 1. ldstMBarrierTokenMap maps mbarrier_init,
+    // mbarrier_inval, and cpAsynBulk to the same mbarrier token.
+    bool mbarrier_token_exists =
+        GpuLower::current()->ldstMBarrierTokenMap().count(expr) != 0;
+
+    bool is_ignorable_tma_smem_alloc =
+        (GpuLower::current()->mBarrierTokenSmemAllocSet().count(expr) != 0);
+
+    bool is_ignorable_mbarrier_init =
+        (expr->isA<kir::MBarrierInit>() && mbarrier_token_exists);
+
+    bool is_ignorable_mbarrier_inval =
+        (expr->isA<kir::MBarrierInvalidate>() && mbarrier_token_exists);
```

Comment:
Why do we want to ignore mbarrier_init and inval? Is it because, before the circular buffering pass, each TMA instruction is created as:
```
init
arriveTX
cpAsyncBulk
wait
inval
```
?
Instead of creating these useless `init` and `inval` and ignore them here and create a new `init` and `inval`, why not just avoid creating these ignorable `init` and `inval`?

Reply:
`init` and `inval` are used in the alias-memory pass to ensure that the lifetime of those shared memory tensors are tracked.

Reply:
How about https://github.com/NVIDIA/Fuser/pull/2833#discussion_r1774111147

Reply:
Push changes to initialize and invalidate mbarriers in allocation pass.